### PR TITLE
Updated CoreStatus enum

### LINKED ISSUE #81: Refactor the CoreStatus enum
- The new statuses are
  - CoreStatus:undefined
  - CoreStatus:terminated
  - CoreStatus:created
  - CoreStatus:opened
  - CoreStatus:precommitted
- Consensus contract should not inherit CoreStatus anymore

### PR DESCRIPTION
**Pull request by gulshanvasnani**
PR contains changes to :
1. Add **terminated** status in CoreStatus enum. Removed **halted** and **corrupted**
2. Change **creation** state to **created** state.
3. Update unit test cases.

fixes #81

### CODE REVIEW

File: contracts/core/Core.sol [hunk 1]
```diff
@@ -678,14 +678,14 @@ contract Core is MasterCopyNonUpgradable, ConsensusModule, MosaicVersion, CoreSt
     }
 
     /**
-     * @notice Joins a validator while core is in creation state.
+     * @notice Joins a validator while core is in created state.
      *         Once the minimum number of validators joined, core is opened,
```
Comment (abhayks1) [inline contracts/core/Core.sol]:
```suggestion
     *         Once the minimum number of validators are joined, core is opened,
```

File: contracts/core/Core.sol [hunk 2]
```diff
@@ -678,14 +678,14 @@ contract Core is MasterCopyNonUpgradable, ConsensusModule, MosaicVersion, CoreSt
...
      *         with the metablock parameters specified in the constructor
```
Comment (abhayks1) [inline contracts/core/Core.sol]:
```suggestion
     *         with the metablock parameters specified in the constructor.
```

File: contracts/core/Core.sol [hunk 3]
```diff
@@ -678,14 +678,14 @@ contract Core is MasterCopyNonUpgradable, ConsensusModule, MosaicVersion, CoreSt
...
-     *         and joined validators during creation.
+     *         and joined validators during core is in created state.
```
Comment (abhayks1) [inline contracts/core/Core.sol]:
```suggestion
     *         Validators join when core is in created state.
```

File: contracts/core/CoreStatusEnum.sol [hunk 4]
```diff
@@ -7,12 +7,10 @@ contract CoreStatusEnum {
     enum CoreStatus {
```
Comment (abhayks1) [inline contracts/core/CoreStatusEnum.sol:7]: Also remove CoreStatusEnum.sol  import from Consensus.sol.

File: contracts/core/Core.sol (hunk 1 above)
Comment (gulshanvasnani) [reply contracts/core/Core.sol]: **Once the minimum number of validators have joined.**
Let me know what you think on suggestion ?

**Comment by abhayks1** [comment]
It will be good if other reviewer also looks into the PR since there were multiple discussions on Core status.

File: contracts/core/Core.sol [hunk 5]
```diff
@@ -685,14 +685,14 @@ contract Core is MasterCopyNonUpgradable, ConsensusModule, MosaicVersion, CoreSt
     }
 
     /**
-     * @notice Joins a validator while core is in creation state.
-     *         Once the minimum number of validators joined, core is opened,
-     *         with the metablock parameters specified in the constructor
-     *         and joined validators during creation.
+     * @notice Joins a validator while core is in created state.
```
Comment (0xsarvesh) [inline contracts/core/Core.sol]:
```suggestion
     * @notice Joins a validator while core is created.
```

File: contracts/core/Core.sol [hunk 6]
```diff
@@ -685,14 +685,14 @@ contract Core is MasterCopyNonUpgradable, ConsensusModule, MosaicVersion, CoreSt
...
+     *         Once the minimum number of validators has joined, core is opened,
+     *         with the metablock parameters specified in the constructor.
+     *         Validators join when core is in created state.
      *
      * @dev Function requires:
      *          - only consensus can call
-     *          - core is in creation state
+     *          - core is in created state
```
Comment (0xsarvesh) [inline contracts/core/Core.sol]:
```suggestion
     *          - core is created and not open.
```

**Review by 0xsarvesh — COMMENTED** [review]
Looks good. 
Some naming convention can be improved. 
joinDuringCreation -> joinBeforeOpen
duringCreation -> beforeOpen

Same naming changes in the documentation.

File: contracts/core/Core.sol (hunk 5 above)
Comment (gulshanvasnani) [reply contracts/core/Core.sol]: Done

File: contracts/core/Core.sol (hunk 6 above)
Comment (gulshanvasnani) [reply contracts/core/Core.sol]: Done

**Review by 0xsarvesh — APPROVED** [review]
Looks good 👍

**Comment by 0xsarvesh** [comment]
Existing unit tests that are modified with this PR for method `joinBeforeOpen` the core contract used is `MockCore`. It's confusing that the unit test is testing methods of Core and the contract name is `MockCore`. To me `MockCore` -> `CoreTest`. `***Test` convention is used for other contracts also for e.g. `ConsensusTest`.

**Pull request merged**